### PR TITLE
OC-1441 Detach step elements from DOM or stop playback of closed steps

### PR DESCRIPTION
**Pull request by jbzdak**
This is initial implementation of OC-1441, this stops playback of videos from previous steps. 

## Course set up

What you need is a course where on a step you have a video, and then you can to the next step (you should still hear the video on that next step --- which is the error). 

1. Create a course with two sections, each containing a single subsection and a single unit. 
2. Each unit contains a step-builder with three steps. First step contains a video and a question, second contains a question, and the last one is the review step. 
3. In one of the subsections use a normal Video XBlock, in second use HTML with OOyala player (OOyala set-up is described in the OC-1441 issue). 

Essentially you should have a course that looks like this: 
```
* Section 1
 * Subsection 1
   * Unit 1
    * Step Builder
     * Step 1
      * Video XBlock with a video
      * Question 
    * Step 2 
     * Question
    * Step 3
     * Review
* Section 2
 * Subsection 2
   * Unit 1
    * Step Builder
     * Step 1
      * HTML Xblock with OOyala video
      * Question 
    * Step 2 
     * Question
    * Step 3
     * Review
```

##  Verification instructions

1. Download the course attached to OC-1441 (for anyone from outside OpenCraft: this course just contains two StepBuilder units containing first step with a video, then step with a quiz and then step with a review; first block contains an Youtube video loaded using a Video XBlock, and second one contains an OOlaya video embedded using HTML Xblock). 
2. Please note that OOyala video needs to be manually started by pasting a snippet to a javascript console --- snippet is attached above the video, and looks like that: `OO.ready(function() { OO.Player.create('xxx', 'yyy_zzz'); });`. After you start the video it should work as it was started "normally" so tests are still valid. 
3. Go to video and block, start the video and wait until it loads, fill in the quiz and go to next step. Note that you still hear video music in the background. 
4. Check the same with the OOyala video subsection. 

## Testing instructions

1. Checkout this branch.
2. Go to video and block, start the video and wait until it loads, fill in the quiz and go to next step. Note that the video is no longer playing in the background. 
3. Fill in rest of the quiz, watch the review step, and go back to the video step. Verify that video is still working properly. 
4. Check the same with the OOyala video subsection.

### CODE REVIEW

File: problem_builder/public/js/mentoring_with_steps.js [hunk 1]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+     * Returns wrapper for active step
+     * @returns {*}, an object containing xblock, and additional metadata (see registerStep where this object is created
+     * for properties)
+     */
+    function getWrapperForActiveStep() {
+        return steps[activeStepIndex];
+    }
+
+    /**
+     * Returns active step)
```
Comment (jbzdak) [inline problem_builder/public/js/mentoring_with_steps.js]: There is a dangling `)` on the end of comment.

File: problem_builder/public/js/mentoring_with_steps.js [hunk 2]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
     var attemptsTemplate = _.template($('#xblock-attempts-template').html());
     var message = $('.sb-step-message', element);
     var checkmark, submitDOM, nextDOM, reviewButtonDOM, tryAgainDOM,
         gradeDOM, attemptsDOM, reviewLinkDOM, submitXHR;
     var reviewStepDOM = $("div.xblock[data-block-type=sb-review-step], div.xblock-v1[data-block-type=sb-review-step]", element);
     var hasAReviewStep = reviewStepDOM.length == 1;
+    
+    /**
+     * Registers step to in this builder, it stores the xblock element, and some additional metadata used to safely
+     * remove and re-attach xblock to the DOM. See showStep and hideStep to see why we need to do it.
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak "Registers step with this StepBuilder instance: Creates and stores a wrapper that contains the XBlock instance and associated DOM element, as well as some additional metadata ..."

Very happy to see documentation for all of the new functions! :)

File: problem_builder/public/js/mentoring_with_steps.js [hunk 3]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+     * @param step xblock instance to register
+     */
+    function registerStep(step) {
+        var $element = $(step.element);
+        var $anchor = $('<span class="xblock-sb-anchor"/>');
+        $anchor.insertBefore($element);
+        var step_wrapper = {
+            idx: step.length,
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak I think you can remove this property, you don't seem to be using it anywhere.

File: problem_builder/public/js/mentoring_with_steps.js [hunk 4]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+            xblock: step,
+            $anchor: $anchor
+        };
+        steps.push(step_wrapper);
+    }
+
+    /**
+     * Returns wrapper for active step
+     * @returns {*}, an object containing xblock, and additional metadata (see registerStep where this object is created
+     * for properties)
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak "an object containing the XBlock instance and associated DOM element for the active step, as well as additional metadata (see registerStep ...)"

File: problem_builder/public/js/mentoring_with_steps.js (hunk 1 above)
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak Leftover ")"

File: problem_builder/public/js/mentoring_with_steps.js [hunk 5]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+     * Returns active step)
+     * @returns {*}
+     */
+    function getActiveStep() {
+        return getWrapperForActiveStep().xblock;
+    }
+
+    /**
+     * Calls a function for each registered step. Object passed to this function is an step wrapper object
+     * (see registerStep where this object is created a list of properties)
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak "Object passed to this function is an" ⟶ "The object passed to this function is a"

"a list of properties" ⟶ "for a list of properties"

File: problem_builder/public/js/mentoring_with_steps.js [hunk 6]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+     *
+     * @param func single arg function.
+     */
+    function forEachStep(func){
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak I like this function, it enables some nice refactorings :)

File: problem_builder/public/js/mentoring_with_steps.js [hunk 7]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+        for (var idx=0; idx<steps.length; idx++) {
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak Nit: Maybe put spaces around `=` and `<`?

File: problem_builder/public/js/mentoring_with_steps.js [hunk 8]
```diff
@@ -7,30 +7,118 @@ function MentoringWithStepsBlock(runtime, element) {
...
+
+    /**
+     * Hides a step
+     * @param step_wrapper
+     */
+    function hideStep(step_wrapper) {
+        // This is a hacky workaround, but it works. It detaches this element from dom, and re-attaches it immediately,
+        // which has the side effect of stopping any video elements (tested with Ooyala and Youtube). 
+        // This solution was chosen as permanently detaching xblocks from DOM is not anticipated by lot of software 
+        // and some code in this repository.
```
Comment (itsjeyd) [inline problem_builder/public/js/mentoring_with_steps.js]: @jbzdak "from dom, and" ⟶ "from the DOM and"

What do you mean by "lot of software"? Might be worth specifying which software :)

**Comment by jbzdak** [comment]
@itsjeyd All review comments have been addressed (fixes should be visible on both PRs).

**Comment by itsjeyd** [comment]
@jbzdak :+1: if YouTube issues can be ignored for now.

Also, we'll need to decide which approach we want to go with (this one or the one from #114) before merging.

**Comment by jbzdak** [comment]
I'm closing this one, I will post a follow up shortly.

**Pull request closed without merging**